### ISSUE TITLE
Project has many hlint warnings

### ISSUE THREAD

**Issue by srid** [nammayatri/nammayatri#31]
As part of #29 I noticed that the project has *a lot* of hlint warnings (using hlint 3.4.1). But somehow the docker image build (which runs hlint) succeeds; is it because of some local hlint configuration not part of the repo?

Here's a sample of warnings:

```
/Users/srid/code/nammayatri/nammayatri/Backend/app/provider-platform/static-offer-driver-app/Main/src/Tools/Notifications.hs:(57,5)-(58,30): Warning: Eta reduce
Found:
  body text = FCMNotificationBody text
Perhaps:
  body = FCMNotificationBody

/Users/srid/code/nammayatri/nammayatri/Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/CustomerSupport.hs:143:36-60: Warning: Use min
Found:
  if x <= 10 then x else 10
Perhaps:
  min x 10

/Users/srid/code/nammayatri/nammayatri/Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Search/Common.hs:69:35-108: Warning: Use min
Found:
  minimum
    [fromInteger searchRequestExpiry, maximum [minExpiry, timeToRide]]
Perhaps:
  min
    (fromInteger searchRequestExpiry) (maximum [minExpiry, timeToRide])
```

**Comment by srid** [comment]
I also saw unnecessary warnings for record-dot-preprocessor generated code, which is an instance of https://github.com/ndmitchell/record-dot-preprocessor/issues/37

**Comment by NikithShetty** [comment]
@srid this issue is resolved in this [PR](https://github.com/nammayatri/nammayatri/pull/46)

`.hlint.yaml` file was missing, we added it back

**Comment by srid** [comment]
@NikithShetty I can confirm that this PR's `.hlint.yaml` does get rid of the warnings around record-dot-preprocessor generated code. However, other warnings (which this issue is about) do remain. Here's the full list of those 32 warnings:

<details>
<summary>hlint output</summary>
<pre>
Backend/app/dashboard/Lib/src/Domain/Types/Person/Type.hs:60:3-27: Warning: Eta reduce
Found:
  fromUnencrypted a = fst a
Perhaps:
  fromUnencrypted = fst

Backend/app/kafka-consumers/src/Consumer/AvailabilityTime/Storage/Queries.hs:71:19-40: Warning: Use max
Found:
  if a > b then a else b
Perhaps:
  max a b

Backend/app/mocks/public-transport-provider-platform/src/Redis.hs:41:3-26: Warning: Eta reduce
Found:
  getKey key = Hed.get key
Perhaps:
  getKey = Hed.get

Backend/app/mocks/public-transport-provider-platform/src/Redis.hs:42:3-34: Warning: Eta reduce
Found:
  setKey key val = Hed.set key val
Perhaps:
  setKey = Hed.set

Backend/app/mocks/public-transport-provider-platform/src/Redis.hs:43:3-26: Warning: Eta reduce
Found:
  delKey key = Hed.del key
Perhaps:
  delKey = Hed.del

Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/Person.hs:121:3-27: Warning: Eta reduce
Found:
  fromUnencrypted a = fst a
Perhaps:
  fromUnencrypted = fst

Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/RideDetails.hs:54:3-27: Warning: Eta reduce
Found:
  fromUnencrypted a = fst a
Perhaps:
  fromUnencrypted = fst

Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/RiderDetails.hs:56:3-27: Warning: Eta reduce
Found:
  fromUnencrypted a = fst a
Perhaps:
  fromUnencrypted = fst

Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/DriverOnboarding/DriverLicense.hs:51:37-57: Suggestion: Evaluate
Found:
  (, salt) licenseNumber
Perhaps:
  (licenseNumber, salt)

Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/DriverOnboarding/DriverLicense.hs:60:3-27: Warning: Eta reduce
Found:
  fromUnencrypted a = fst a
Perhaps:
  fromUnencrypted = fst

Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/DriverOnboarding/IdfyVerification.hs:54:38-59: Suggestion: Evaluate
Found:
  (, salt) documentNumber
Perhaps:
  (documentNumber, salt)

Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/DriverOnboarding/IdfyVerification.hs:63:3-27: Warning: Eta reduce
Found:
  fromUnencrypted a = fst a
Perhaps:
  fromUnencrypted = fst

Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/DriverOnboarding/VehicleRegistrationCertificate.hs:56:41-65: Suggestion: Evaluate
Found:
  (, salt) certificateNumber
Perhaps:
  (certificateNumber, salt)

Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/DriverOnboarding/VehicleRegistrationCertificate.hs:65:3-27: Warning: Eta reduce
Found:
  fromUnencrypted a = fst a
Perhaps:
  fromUnencrypted = fst

Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/Merchant/MerchantServiceConfig.hs:(40,3)-(57,8): Warning: Eta reduce
Found:
  readsPrec d' r'
    = readParen
        (d' > app_prec)
        (\ r
           -> [(MapsService v1, r2) |
                 r1 <- stripPrefix "Maps_" r,
                 (v1, r2) <- readsPrec (app_prec + 1) r1]
                ++
                  [(SmsService v1, r2) |
                     r1 <- stripPrefix "Sms_" r,
                     (v1, r2) <- readsPrec (app_prec + 1) r1]
                    ++
                      [(WhatsappService v1, r2) |
                         r1 <- stripPrefix "Whatsapp_" r,
                         (v1, r2) <- readsPrec (app_prec + 1) r1])
        r'
Perhaps:
  readsPrec d'
    = readParen
        (d' > app_prec)
        (\ r
           -> [(MapsService v1, r2) |
                 r1 <- stripPrefix "Maps_" r,
                 (v1, r2) <- readsPrec (app_prec + 1) r1]
                ++
                  [(SmsService v1, r2) |
                     r1 <- stripPrefix "Sms_" r,
                     (v1, r2) <- readsPrec (app_prec + 1) r1]
                    ++
                      [(WhatsappService v1, r2) |
                         r1 <- stripPrefix "Whatsapp_" r,
                         (v1, r2) <- readsPrec (app_prec + 1) r1])

Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/DriverOnboarding.hs:34:29-54: Warning: Use mapMaybe
Found:
  catMaybes $ toMsg <$> errs
Perhaps:
  mapMaybe toMsg errs

Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool.hs:292:5-65: Warning: Eta reduce
Found:
  getRatios fn arr = mapM (\ dId -> (dId . getId,) <$> fn dId) arr
Perhaps:
  getRatios fn = mapM (\ dId -> (dId . getId,) <$> fn dId)

Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Tools/Notifications.hs:(106,5)-(107,30): Warning: Eta reduce
Found:
  body text = FCMNotificationBody text
Perhaps:
  body = FCMNotificationBody

Backend/app/provider-platform/static-offer-driver-app/Main/src/Domain/Action/Beckn/Search.hs:110:31-96: Warning: Use min
Found:
  minimum [fromInteger caseExpiry_, maximum [minExpiry, timeToRide]]
Perhaps:
  min (fromInteger caseExpiry_) (maximum [minExpiry, timeToRide])

Backend/app/provider-platform/static-offer-driver-app/Main/src/Domain/Action/Beckn/Search.hs:110:65-95: Warning: Use max
Found:
  maximum [minExpiry, timeToRide]
Perhaps:
  max minExpiry timeToRide

Backend/app/provider-platform/static-offer-driver-app/Main/src/Domain/Types/Person.hs:116:3-27: Warning: Eta reduce
Found:
  fromUnencrypted a = fst a
Perhaps:
  fromUnencrypted = fst

Backend/app/provider-platform/static-offer-driver-app/Main/src/Domain/Types/RideDetails.hs:54:3-27: Warning: Eta reduce
Found:
  fromUnencrypted a = fst a
Perhaps:
  fromUnencrypted = fst

Backend/app/provider-platform/static-offer-driver-app/Main/src/Domain/Types/RiderDetails.hs:49:3-27: Warning: Eta reduce
Found:
  fromUnencrypted a = fst a
Perhaps:
  fromUnencrypted = fst

Backend/app/provider-platform/static-offer-driver-app/Main/src/Domain/Types/Merchant/MerchantServiceConfig.hs:(38,3)-(51,8): Warning: Eta reduce
Found:
  readsPrec d' r'
    = readParen
        (d' > app_prec)
        (\ r
           -> [(MapsService v1, r2) |
                 r1 <- stripPrefix "Maps_" r,
                 (v1, r2) <- readsPrec (app_prec + 1) r1]
                ++
                  [(SmsService v1, r2) |
                     r1 <- stripPrefix "Sms_" r,
                     (v1, r2) <- readsPrec (app_prec + 1) r1])
        r'
Perhaps:
  readsPrec d'
    = readParen
        (d' > app_prec)
        (\ r
           -> [(MapsService v1, r2) |
                 r1 <- stripPrefix "Maps_" r,
                 (v1, r2) <- readsPrec (app_prec + 1) r1]
                ++
                  [(SmsService v1, r2) |
                     r1 <- stripPrefix "Sms_" r,
                     (v1, r2) <- readsPrec (app_prec + 1) r1])

Backend/app/provider-platform/static-offer-driver-app/Main/src/Tools/Notifications.hs:(57,5)-(58,30): Warning: Eta reduce
Found:
  body text = FCMNotificationBody text
Perhaps:
  body = FCMNotificationBody

Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/CustomerSupport.hs:143:36-60: Warning: Use min
Found:
  if x <= 10 then x else 10
Perhaps:
  min x 10

Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Search/Common.hs:70:35-108: Warning: Use min
Found:
  minimum
    [fromInteger searchRequestExpiry, maximum [minExpiry, timeToRide]]
Perhaps:
  min
    (fromInteger searchRequestExpiry) (maximum [minExpiry, timeToRide])

Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Search/Common.hs:70:77-107: Warning: Use max
Found:
  maximum [minExpiry, timeToRide]
Perhaps:
  max minExpiry timeToRide

Backend/app/rider-platform/rider-app/Main/src/Domain/Types/Person.hs:127:3-27: Warning: Eta reduce
Found:
  fromUnencrypted a = fst a
Perhaps:
  fromUnencrypted = fst

Backend/app/rider-platform/rider-app/Main/src/Domain/Types/Merchant/MerchantServiceConfig.hs:(40,3)-(57,8): Warning: Eta reduce
Found:
  readsPrec d' r'
    = readParen
        (d' > app_prec)
        (\ r
           -> [(MapsService v1, r2) |
                 r1 <- stripPrefix "Maps_" r,
                 (v1, r2) <- readsPrec (app_prec + 1) r1]
                ++
                  [(SmsService v1, r2) |
                     r1 <- stripPrefix "Sms_" r,
                     (v1, r2) <- readsPrec (app_prec + 1) r1]
                    ++
                      [(WhatsappService v1, r2) |
                         r1 <- stripPrefix "Whatsapp_" r,
                         (v1, r2) <- readsPrec (app_prec + 1) r1])
        r'
Perhaps:
  readsPrec d'
    = readParen
        (d' > app_prec)
        (\ r
           -> [(MapsService v1, r2) |
                 r1 <- stripPrefix "Maps_" r,
                 (v1, r2) <- readsPrec (app_prec + 1) r1]
                ++
                  [(SmsService v1, r2) |
                     r1 <- stripPrefix "Sms_" r,
                     (v1, r2) <- readsPrec (app_prec + 1) r1]
                    ++
                      [(WhatsappService v1, r2) |
                         r1 <- stripPrefix "Whatsapp_" r,
                         (v1, r2) <- readsPrec (app_prec + 1) r1])

Backend/lib/beckn-spec/src/Beckn/Types/Core/Taxi/Common/ItemCode.hs:(52,3)-(77,8): Warning: Eta reduce
Found:
  readsPrec d' r'
    = readParen
        (d' > app_prec)
        (\ r
           -> [(ItemCode ONE_WAY_TRIP v1 Nothing Nothing, r2) |
                 r1 <- stripPrefix "ONE_WAY_TRIP_" r,
                 (v1, r2) <- readsPrec (app_prec + 1) r1]
                ++
                  [(ItemCode RENTAL_TRIP v1 (Just v2) (Just v3), r6) |
                     r1 <- stripPrefix "RENTAL_TRIP_" r,
                     (v1, r2) <- parseValueAtStart (app_prec + 1) r1,
                     r3 <- stripPrefix "_" r2,
                     (v2, r4) <- parseValueAtStart (app_prec + 1) r3,
                     r5 <- stripPrefix "_" r4,
                     (v3, r6) <- readsPrec (app_prec + 1) r5]
                    ++
                      [(ItemCode DRIVER_OFFER_ESTIMATE v1 Nothing Nothing, r2) |
                         r1 <- stripPrefix "DRIVER_OFFER_ESTIMATE_" r,
                         (v1, r2) <- readsPrec (app_prec + 1) r1]
                        ++
                          [(ItemCode DRIVER_OFFER v1 Nothing Nothing, r2) |
                             r1 <- stripPrefix "DRIVER_OFFER_" r,
                             (v1, r2) <- readsPrec (app_prec + 1) r1])
        r'
Perhaps:
  readsPrec d'
    = readParen
        (d' > app_prec)
        (\ r
           -> [(ItemCode ONE_WAY_TRIP v1 Nothing Nothing, r2) |
                 r1 <- stripPrefix "ONE_WAY_TRIP_" r,
                 (v1, r2) <- readsPrec (app_prec + 1) r1]
                ++
                  [(ItemCode RENTAL_TRIP v1 (Just v2) (Just v3), r6) |
                     r1 <- stripPrefix "RENTAL_TRIP_" r,
                     (v1, r2) <- parseValueAtStart (app_prec + 1) r1,
                     r3 <- stripPrefix "_" r2,
                     (v2, r4) <- parseValueAtStart (app_prec + 1) r3,
                     r5 <- stripPrefix "_" r4,
                     (v3, r6) <- readsPrec (app_prec + 1) r5]
                    ++
                      [(ItemCode DRIVER_OFFER_ESTIMATE v1 Nothing Nothing, r2) |
                         r1 <- stripPrefix "DRIVER_OFFER_ESTIMATE_" r,
                         (v1, r2) <- readsPrec (app_prec + 1) r1]
                        ++
                          [(ItemCode DRIVER_OFFER v1 Nothing Nothing, r2) |
                             r1 <- stripPrefix "DRIVER_OFFER_" r,
                             (v1, r2) <- readsPrec (app_prec + 1) r1])

Backend/test/src/Mobility/Transporter/LocationUpdates.hs:80:5-100: Warning: Eta reduce
Found:
  makePointsOutdated points
    = fmap
        (\ point -> point {ts = addUTCTime (negate 600) point . ts}) points
Perhaps:
  makePointsOutdated
    = fmap (\ point -> point {ts = addUTCTime (negate 600) point . ts})

32 hints

</pre>
</details>

Note: I'm using hlint v3.4.1 from nixpkgs.

**Comment by utkarsh911** [comment]
Yes, that is because of version mismatch. Also, these warnings looks valid to me. We may apply these.

**Comment by 0utkarsh** [comment]
Yes, that is because of version mismatch. Also, these warnings looks valid to me. We may apply these.

**Comment by srid** [comment]
When we switch from stack to Nix, we'll naturally upgrade to hlint 3.4.1 (unless you guys want to strictly stay with 3.2.7).

If you choose to upgrade, the hlint warnings can also be dealt with prior to Nix migration. [LTS 20.11](https://www.stackage.org/lts-20.11) has hlint 3.4.1 ... so, IIUC, changing this line should be enough to get these warnings showing in the image build:

https://github.com/nammayatri/nammayatri/blob/3fd4e24488cfb95c3a260ce7e514cb4e321e86a2/Backend/Dockerfile.dep#L21-L22
